src/ContainerGetCapableTrait.php

```diff
@@ -32,50 +23,77 @@
      *
      * @since [*next-version*]
      *
-     * @param array|stdClass|ContainerInterface $container The container to read from.
-     * @param string|Stringable                 $key       The key of hte value to retrieve.
-     *
-     * @return mixed The value mapped to the given key.
+     * @param array|ArrayAccess|stdClass|ContainerInterface $container The container to read from.
+     * @param string|int|float|bool|Stringable              $key       The key of hte value to retrieve.
      *
      * @throws ContainerExceptionInterface If an error occurred while reading from the container.
      * @throws NotFoundExceptionInterface  If the key was not found in the container.
+     *
+     * @return mixed The value mapped to the given key.
      */
     protected function _containerGet($container, $key)
     {
-        $key = $this->_normalizeString($key);
-        $isContainer = $container instanceof ContainerInterface;
-        $isArrayLike = is_array($container) || $container instanceof ArrayAccess;
-        $isObject = $container instanceof stdClass;
-
-        if (!$isContainer && !$isArrayLike && !$isObject) {
-            throw $this->_createInvalidArgumentException(
-                $this->__('Argument #1 is not a valid container'),
-                null,
-                null,
-                $container
-            );
+        $origKey = $key;
+        $key     = $this->_normalizeString($key);
+        // NotFoundExceptionInterface#getDataKey() returns `string` or `Stringable`,
+        // so normalize only other types, and preserve original
+        $origKey = is_string($key) || $key instanceof Stringable
```

Comment:
The `$key` is being normalized to a string (_line 37_).  
This means all scalar values will become strings and `Stringable` instances will be reduced to strings. In other words, all non-null values of `$key` result in `$key` becoming a string.

As such, this condition will never fail for non-null values, making `$origKey` always equal itself.

Is this intended? Or was there some other intended purpose of this check?

Reply:
Didn't quite understand. `$origKey` will equal itself if it is a `string` or `Stringable`. Everything else gets normalized. Or am I missing something?

Reply:
````

Reply:
````

Reply:
```
// origKey | value: unknown | type: mixed
// key     | value: unknown | type: mixed
$origKey = key;

// origKey | value: unknown | type: mixed
// key     | value: unknown | type: string
$key = $this->_normalizeString($key);

// origKey | value: unknown | type: mixed
// key     | value: unknown | TYPE: STRING
$origKey = is_string($key) || $key instanceof Stringable
    ? $origKey // This will always happen.
    : $key;    // This will never happen.
```

Reply:
Good point.